### PR TITLE
store/tikv/region_cache: Backoff when region epoch is ahead of TiKV

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

### What problem does this PR solve? 

This PR is part of the solution that when TiKV is merging two regions, a request to the source region may be directed to the target region but the target region haven't finished merging, then the requested key is out of the target region's range, causing TiKV panics.

### What is changed and how it works?

TiKV will strictly check region epoch, after @overvenus submitting a PR to do that. StaleEpoch error will be renamed to EpochNotMatch, and if the request's epoch is ahead of TiKV's, TiKV will still send the older region meta back to TiDB. Then if TiDB founds that TiKV sends back a older region meta, it will backoff.

### Check List 

Tests 

 - Unit tests

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - None

### CODE REVIEW

File: store/mockstore/mocktikv/rpc.go [hunk 1]
```diff
@@ -171,14 +171,14 @@ func (h *rpcHandler) checkRequestContext(ctx *kvrpcpb.Context) *errorpb.Error {
...
 	if !proto.Equal(region.GetRegionEpoch(), ctx.GetRegionEpoch()) {
 		nextRegion, _ := h.cluster.GetRegionByKey(region.GetEndKey())
-		newRegions := []*metapb.Region{region}
+		currentRegions := []*metapb.Region{region}
 		if nextRegion != nil {
-			newRegions = append(newRegions, nextRegion)
+			currentRegions = append(currentRegions, nextRegion)
 		}
 		return &errorpb.Error{
 			Message: *proto.String("stale epoch"),
```
Comment (zhangjinpeng1987) [inline store/mockstore/mocktikv/rpc.go]: s/stale epoch/epoch not match

**Comment by zhangjinpeng1987** [comment]
@disksing @coocood PTAL

**Comment by MyonKeminta** [comment]
@disksing @tiancaiamao If you have time these days, please kindly help me review this PR, thank you very much!

**Comment by tiancaiamao** [comment]
> when region epoch is ahead of TiKV

I wonder how could this happen?

File: store/tikv/region_cache_test.go [hunk 2]
```diff
@@ -299,6 +299,28 @@ func (s *testRegionCacheSuite) TestRequestFail2(c *C) {
...
+	r1 := metapb.Region{Id: 1, RegionEpoch: &metapb.RegionEpoch{Version: 9, ConfVer: 10}}
+	r2 := metapb.Region{Id: 1, RegionEpoch: &metapb.RegionEpoch{Version: 10, ConfVer: 9}}
+
+	bo := NewBackoffer(context.Background(), 2000000)
+
+	err := cache.OnRegionEpochNotMatch(bo, &RPCContext{Region: region.VerID()}, []*metapb.Region{&r1})
+	c.Assert(err, IsNil)
+	err = cache.OnRegionEpochNotMatch(bo, &RPCContext{Region: region.VerID()}, []*metapb.Region{&r2})
+	c.Assert(err, IsNil)
+	c.Assert(len(bo.errors), Equals, 2)
```
Comment (tiancaiamao) [inline store/tikv/region_cache_test.go]: Why the errors length is 2 ?

Comment (MyonKeminta) [reply store/tikv/region_cache_test.go]: One should occur at L317 and another should occur at L319.

**Comment by disksing** [comment]
LGTM.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**